packages/context/src/resolver.ts

```diff
@@ -56,51 +56,20 @@ export function instantiateClass<T>(
   }
   const argsOrPromise = resolveInjectedArguments(ctor, '', ctx, session);
   const propertiesOrPromise = resolveInjectedProperties(ctor, ctx, session);
-  let inst: ValueOrPromise<T>;
-  if (isPromiseLike(argsOrPromise)) {
-    // Instantiate the class asynchronously
-    inst = argsOrPromise.then(args => {
-      /* istanbul ignore if */
-      if (debug.enabled) {
-        debug('Injected arguments for %s():', ctor.name, args);
-      }
-      return new ctor(...args);
-    });
-  } else {
+  const inst: ValueOrPromise<T> = resolveValueOrPromise(argsOrPromise, args => {
     /* istanbul ignore if */
     if (debug.enabled) {
-      debug('Injected arguments for %s():', ctor.name, argsOrPromise);
+      debug('Injected arguments for %s():', ctor.name, args);
     }
-    // Instantiate the class synchronously
-    inst = new ctor(...argsOrPromise);
-  }
-  if (isPromiseLike(propertiesOrPromise)) {
-    return propertiesOrPromise.then(props => {
-      /* istanbul ignore if */
-      if (debug.enabled) {
-        debug('Injected properties for %s:', ctor.name, props);
-      }
-      if (isPromiseLike(inst)) {
-        // Inject the properties asynchronously
-        return inst.then(obj => Object.assign(obj, props));
-      } else {
-        // Inject the properties synchronously
-        return Object.assign(inst, props);
-      }
-    });
-  } else {
-    if (isPromiseLike(inst)) {
-      /* istanbul ignore if */
-      if (debug.enabled) {
-        debug('Injected properties for %s:', ctor.name, propertiesOrPromise);
-      }
-      // Inject the properties asynchronously
-      return inst.then(obj => Object.assign(obj, propertiesOrPromise));
-    } else {
-      // Inject the properties synchronously
-      return Object.assign(inst, propertiesOrPromise);
+    return new ctor(...args);
+  });
+  return resolveValueOrPromise(propertiesOrPromise, props => {
```

Comment:
is there a reason as to why we don't use a promise style here?

Reply:
We need to do the work synchronously or asynchronously based on the return value. The usage of `ValueOrPromise` to allow us to call `context.getSync()` and `context.get()`.